src/Core/FakeVimCommand.cpp

```diff
@@ -0,0 +1,263 @@
+            lang = "Python";
+        else if (args == "java")
+            lang = "Java";
+        else if (args.isEmpty())
+            lang = SettingsHelper::getDefaultLanguage();
+
+        if (!lang.isEmpty())
+            appwin->openTab("", lang);
+        else
+            showError(tr("new requires empty or one of 'cpp', 'java' and 'python' argument, got %1").arg(args));
+        break;
+    }
+
+    case CommandTypes::NEXT_TAB: {
+        auto total = appwin->totalTabs();
+        auto curr = appwin->indexOfWindow(appwin->currentWindow());
+        int next = (curr + 1) % total;
+        if (next != curr)
+            appwin->setTabAt(next);
+        break;
+    }
+    case CommandTypes::LAST_TAB: {
+        auto total = appwin->totalTabs();
+        auto curr = appwin->indexOfWindow(appwin->currentWindow());
+        int last = curr ? curr - 1 : total - 1;
+        if (last != curr)
+            appwin->setTabAt(last);
+        break;
+    }
+
+    case CommandTypes::OPEN: {
+        QString path = args;
+        if (path.isEmpty())
+        {
+            showError(tr("open requires a file path"));
+            break;
+        }
+
+        if (path.startsWith('~'))
+            path = QDir::home().absolutePath() + "/" + args.mid(1);
+
+        QFileInfo file(path);
+        if (file.isRelative())
+            file.makeAbsolute();
+
+        if (!Util::cppSuffix.contains(file.suffix()) && !Util::javaSuffix.contains(file.suffix()) &&
+            !Util::javaSuffix.contains(file.suffix()))
+        {
+            showError(tr("%1 is not c++, python or java source file").arg(file.absoluteFilePath()));
+            break;
+        }
+
+        if (!file.exists() && !hasbang)
+        {
+            showError(
+                tr("%1 does not exists. To open non-existing file use open! instead").arg(file.absoluteFilePath()));
+            break;
+        }
+
+        appwin->openTab(file.absoluteFilePath());
+        break;
+    }
+    case CommandTypes::COMPILE: {
+        appwin->on_actionCompile_triggered();
+        break;
+    }
+    case CommandTypes::COMPILE_RUN: {
+        appwin->on_actionCompileRun_triggered();
+        break;
+    }
+    case CommandTypes::RUN: {
+        int ans = -1;
+        if (args.isEmpty())
+            ans = 0;
+        bool ok = false;
+        int caseNum = args.toInt(&ok);
+        if (ok)
+            ans = caseNum;
+        else if (!args.isEmpty())
+        {
+            showError(tr("%1 is not a number").arg(caseNum));
+            break;
+        }
+
+        // Parsed index correctly
+        if (ans == 0)
+            appwin->on_actionRun_triggered();
+        else if (ans > 0 && appwin->currentWindow() && appwin->currentWindow()->testcases &&
+                 appwin->currentWindow()->testcases->count() >= ans)
+            appwin->currentWindow()->runTestCase(ans - 1);
+        else
+            showError(tr("cannot run testcase %1").arg(args));
+        break;
```

Comment:
I think the logic could be clearer:

1. If `args.isEmpty()`, do something and break.
2. If the argument is not a number, show error and break.
3. If the case number is valid, do something and break.
4. Do something.

Instead of:

1. `ans = -1`.
2. If `args.isEmpty()`, `ans = 0`.
3. Else if the case number is a number, `ans = caseNum`; otherwise, show error and break.
4. If `ans == 0`, do something.
5. If `ans > 0` and `ans` is valid, do something.
6. Else do something.

Reply:
```
1. if(args is empty) do stuff
2. else 
       2.1 if(args is not parseable to int) error nan
       2.2 else 
               2.2.1 if (args is in range) runOnly this testcase
               2.2.2 else error not a valid number.  
```

I did it like this, no breaks